internal/proxy/proxy.go

```diff
@@ -192,165 +192,189 @@ func (p *Proxy) Handle(w http.ResponseWriter, r *http.Request) {
 	}).ServeHTTP(w, r)
 }
 
-func (p *Proxy) handleJenkinsUIRequest(w http.ResponseWriter, r *http.Request, requestLogEntry *log.Entry) (cacheKey string, ns string, noProxy bool) {
+func (p *Proxy) handleJenkinsUIRequest(responseWriter http.ResponseWriter, request *http.Request, requestLogEntry *log.Entry) (cacheKey string, namespace string, noProxy bool) {
 	//redirect determines if we need to redirect to auth service
 	needsAuth := true
 	noProxy = true
 
-	//redirectURL is used for auth service as a target of successful auth redirect
-	redirectURL, err := url.ParseRequestURI(fmt.Sprintf("%s%s", strings.TrimRight(p.redirect, "/"), r.URL.Path))
+	// redirectURL is used for auth service as a target of successful auth redirect
+	redirectURL, err := url.ParseRequestURI(fmt.Sprintf("%s%s", strings.TrimRight(p.redirect, "/"), request.URL.Path))
 	if err != nil {
-		p.HandleError(w, err, requestLogEntry)
+		p.HandleError(responseWriter, err, requestLogEntry)
 		return
 	}
 
 	//If the user provides OSO token, we can directly proxy
-	if _, ok := r.Header["Authorization"]; ok { //FIXME Do we need this?
+	if _, ok := request.Header["Authorization"]; ok { //FIXME Do we need this?
 		needsAuth = false
 	}
 
-	if tj, ok := r.URL.Query()["token_json"]; ok { //If there is token_json in query, process it, find user info and login to Jenkins
-		if len(tj) < 1 {
-			p.HandleError(w, errors.New("could not read JWT token from URL"), requestLogEntry)
+	namespace, noProxy = p.processTokenJSON(responseWriter, request, requestLogEntry, redirectURL)
```

Comment:
Ok, now I get it. It is a start, but I don't think there is a need for returning anything here. Especially look at  p.checkCookies. 

I guess as a first step this works, but the next is to eliminate the returns. Thale the last if statement for example `if needsAuth`. This could be pulled out into a function which then can get called.

Also, we tried to remove most of the used naked returns in the code. In most places we have done it already, but the code in `proxy.go` uses still a lot of them. The use of naked returns is really discouraged and should if ever be only used for very small methods. IMO in the current code it just makes it hard to understand. Let's get rid of them.

Reply:
That said, I think we need to keep some conditional there. Pulling the conditionals into the methods obfuscates the code. Now it looks like, you are always processing the JSON token and you are always proecssing the cookies. There are several phases here which are not obvious anymore. 

Leaving the outer if statements here will makes things clearer